nx/lib/nx.ex

```diff
@@ -7835,40 +7836,168 @@ defmodule Nx do
   """
   @doc type: :cumulative
   def cumulative_max(tensor, opts \\ []),
-    do: cumulative_op(tensor, opts, :cumulative_max, :window_max)
+    do: cumulative_op(tensor, opts, :cumulative_max, &Nx.max/2)
 
-  defp cumulative_op(tensor, opts, op, window_op) do
+  defp cumulative_op(tensor, opts, op, reduce_fun) do
     opts = keyword!(opts, axis: 0, reverse: false)
     reverse = opts[:reverse]
     tensor = to_tensor(tensor)
     axis = Nx.Shape.normalize_axis(tensor.shape, opts[:axis], tensor.names)
 
-    Nx.Shared.optional(op, [tensor, [axis: axis, reverse: reverse]], tensor, fn tensor,
-                                                                                axis: axis,
-                                                                                reverse: reverse ->
-      shape = shape(tensor)
-      axis_size = elem(shape, axis)
-      rank = rank(shape)
-
-      padding =
-        if reverse do
-          {0, 0}
-          |> List.duplicate(rank)
-          |> List.replace_at(axis, {0, axis_size - 1})
-        else
-          {0, 0}
-          |> List.duplicate(rank)
-          |> List.replace_at(axis, {axis_size - 1, 0})
-        end
+    Nx.Shared.optional(op, [tensor, [axis: axis, reverse: reverse]], tensor, fn tensor, opts ->
+      associative_scan(tensor, reduce_fun, opts)
+    end)
+  end
 
-      window_shape =
-        1
-        |> List.duplicate(rank)
-        |> List.to_tuple()
-        |> put_elem(axis, axis_size)
+  # Scans the given tensor using an associative binary operator.
+  #
+  # The scanning function must be associative and perform an element-wise
+  # operation over the `:axis` dimension.
+  #
+  # ## Options
+  #
+  #   * `:axis` - the axis to scan along. Defaults to `0`
+  #
+  #   * `:reverse` - whether to scan in the opposite direction. Defaults to `false`
+  #
+  # ## Examples
+  #
+  # A cumulative sum of numbers can be expressed as:
+  #
+  #     iex> Nx.associative_scan(Nx.tensor([1, 2, 3, 4]), &Nx.add/2)
+  #     #Nx.Tensor<
+  #       s64[4]
+  #       [1, 3, 6, 10]
+  #     >
+  #
+  # Or a reversed one:
+  #
+  #     iex> Nx.associative_scan(Nx.tensor([1, 2, 3, 4]), &Nx.add/2, reverse: true)
+  #     #Nx.Tensor<
+  #       s64[4]
+  #       [10, 9, 7, 4]
+  #     >
+  #
+  # A cumulative product of a sequence of matrices:
+  #
+  #     iex> matrices = Nx.tensor([[2, 0], [0, 2]]) |> Nx.tile([3, 1, 1])
+  #     iex> Nx.associative_scan(matrices, &Nx.dot(&1, [2], [0], &2, [1], [0]))
+  #     #Nx.Tensor<
+  #       s64[3][2][2]
+  #       [
+  #         [
+  #           [2, 0],
+  #           [0, 2]
+  #         ],
+  #         [
+  #           [4, 0],
+  #           [0, 4]
+  #         ],
+  #         [
+  #           [8, 0],
+  #           [0, 8]
+  #         ]
+  #       ]
+  #     >
+  #
+  defnp associative_scan(tensor, fun, opts \\ []) do
```

Comment:
We don't want this to be `defnp` because everything in the `Nx` module is considered a `defn` already. 

Otherwise I am worried we will run into weird scenarios like jitting this function while compiling another JIT and so on.

So please remove `import Nx.Defn`, and rewrite those to `defp` and we should be good!